cmd/dra-example-controller/allocations.go

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2023 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"sync"
+
+	nascrd "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/nas/v1alpha1"
+)
+
+type PerNodeAllocatedClaims struct {
+	sync.RWMutex
+	allocations map[string]map[string]nascrd.AllocatedDevices
+}
+
+func NewPerNodeAllocatedClaims() *PerNodeAllocatedClaims {
+	return &PerNodeAllocatedClaims{
+		allocations: make(map[string]map[string]nascrd.AllocatedDevices),
+	}
+}
+
+func (p *PerNodeAllocatedClaims) Exists(claimUID, node string) bool {
+	p.RLock()
+	defer p.RUnlock()
+
+	_, exists := p.allocations[claimUID]
+	if !exists {
+		return false
+	}
+
+	_, exists = p.allocations[claimUID][node]
+	return exists
+}
+
+func (p *PerNodeAllocatedClaims) Get(claimUID, node string) nascrd.AllocatedDevices {
+	p.RLock()
+	defer p.RUnlock()
+
+	if !p.Exists(claimUID, node) {
+		return nascrd.AllocatedDevices{}
+	}
+	return p.allocations[claimUID][node]
+}
+
+func (p *PerNodeAllocatedClaims) VisitNode(node string, visitor func(claimUID string, allocation nascrd.AllocatedDevices)) {
+	p.RLock()
+	for claimUID := range p.allocations {
+		if allocation, exists := p.allocations[claimUID][node]; exists {
+			p.RUnlock()
+			visitor(claimUID, allocation)
+			p.RLock()
+		}
+	}
+	p.RUnlock()
+}
+
+func (p *PerNodeAllocatedClaims) Visit(visitor func(claimUID, node string, allocation nascrd.AllocatedDevices)) {
+	p.RLock()
```

Comment:
Any reason not to use _defer_ here?

Reply:
Just for symmetry, since we unlock and relock across the `visit()` callback.

Reply:
I'd suggest to defer just for symmetry with the rest of the methods and to avoid surprises in future, when the logic may include returns. It's a nit, so up to you.

Reply:
I think I'd rather leave it as is. Normally you use the pattern of lock() --> defer(unlock) when you want to lock across the whole function and make sure you *always* unlock on every return path. In this case I unlock and relock within the body of the function and there is only one return path, so I'd rather have all locking and unlocking points be explicit.